site/js/helpers.js

```diff
@@ -115,11 +107,184 @@ function copyToClipboard(text) {
+            return `selector:${selectorType}`;
     }
+}
+
+function generateRuleSeparatorHTML(ruleSeparator) {
+    return `<span title="${i18next.t('texts.rule separator ' + ruleSeparator)}" class="rule_separator">` +
+           `<a target="_blank" class="specification" href="${specification_url}#section:rule_separators">${ruleSeparator}</a></span><br>`;
+}
+
+function generateSelectorHTML(selectorType, selectorValue) {
+    const fragmentIdentifier = getFragmentIdentifier(selectorType);
+    const translationKey = selectorType.match(/(?:state|comment)/) ? 'modifier' : 'selector';
+
+    return `<span title="${i18next.t(`words.${translationKey}`, { name: selectorType })}" class="${selectorType}">` +
+           `<a target="_blank" class="specification" href="${specification_url}#${fragmentIdentifier}">${selectorValue}</a></span>`;
+}
+
+function generateValueExplanation(prettifiedValueArray) {
+    const parts = [
+        `${i18next.t('texts.prettified value for displaying')}:<br />`,
+        '<p class="value_explanation">'
+    ];
+
+    for (let nrule = 0; nrule < prettifiedValueArray[0].length; nrule++) {
+        if (nrule !== 0) {
+            const ruleSeparator = prettifiedValueArray[1][nrule][1]
+                ? ' ||'
+                : (prettifiedValueArray[1][nrule][0][0][1] === 'rule separator' ? ',' : ';');
+
+            parts.push(generateRuleSeparatorHTML(ruleSeparator));
+        }
+
+        parts.push('<span class="one_rule">');
+
+        const selectors = prettifiedValueArray[0][nrule];
+        for (let nselector = 0; nselector < selectors.length; nselector++) {
+            const selectorType = selectors[nselector][0][2];
+            const selectorValue = selectors[nselector][1];
 
+            parts.push(generateSelectorHTML(selectorType, selectorValue));
+
+            if (nselector + 1 < selectors.length) {
+                parts.push(' ');
+            }
+        }
+
+        parts.push('</span>');
+    }
+
+    parts.push('</p></div>');
+    return parts.join('');
+}
+
+function generateResultsHTML() {
+    return `
+        <div class="matching-rule-card">
+            <div class="status-label">${i18next.t('texts.MatchingRule')}</div>
+            <div class="matching-rule-value" id="matching-rule-display"></div>
+        </div>
+    `;
+}
+
+function handleDiffComparison(oh, diffValue, mode, valueExplanation) {
+    const diffValueElement = document.getElementById('diff_value');
+
+    if (diffValue.length === 0) {
+        diffValueElement.style.backgroundColor = '';
+        return valueExplanation;
+    }
+
+    let isEqualTo;
+    try {
+        isEqualTo = oh.isEqualTo(new opening_hours(diffValue, nominatim, {
+            'mode': mode,
+            'warnings_severity': 7,
+            'locale': i18next.language
+        }));
+    } catch {
+        diffValueElement.style.backgroundColor = evaluation_tool_colors.error;
+        return valueExplanation;
+    }
+
+    if (typeof isEqualTo !== 'object') {
```

Comment:
```suggestion
    if (isEqualTo === null || typeof isEqualTo !== 'object') {
```
Easy to forget. Not sure if this is relevant here